### PR TITLE
Calculations of current-dipole moments and potentials

### PR DESCRIPTION
**Pull request by solveignaess**
issue #7 
- cell.py includes methods for computing axial currents and current dipole moments
- eegmegcalc.py can (for now) compute extracellular potentials based on the current dipole approximation in the inhomogeneous 4-sphere head model. This class should also contain methods for computing extracellular potentials in 1-sphere head model, and in infinite, homogeneous 3D conductor. MEG computations should be added to this file.

### CODE REVIEW

**Comment by espenhgn** [comment]
@solveignaess 
That ExpSyn fail in `test_cell_set_synapse_00` can be fixed by replacing the line 
```self.assertEqual(cell.synlist[0].hname(), 'ExpSyn[0]')```
by
```self.assertTrue('ExpSyn' in cell.synlist[0].hname())```

I will make that change in another working branch but you can also incorporate it

File: LFPy/cell.py
Comment (espenhgn) [inline LFPy/cell.py]: I think it may have been intentional but I propose that the returned arguments are shaped `(2*self.totnsegs, ...)`. Now the returned arguments are always shaped `(2*self.totnsegs-2, ...)`. I know the start-to-mid point axial current of soma section would normally be 0, as well as mid-to-end axial current of terminating segments will be zero. If number of currents are 2*self.totnsegs then one should be able to unambigously access axial currents of a segment, granted that the order of axial currents and cell segments is similar

File: LFPy/cell.py
Comment (espenhgn) [inline LFPy/cell.py]: Also here the description is a bit imprecise, it is the axial resistance from the parent segment, not the midpoint of the parent segment, see the description of `neuron.h.ri`: 

>     Return the resistance (in megohms) between the center of the segment containing x and its parent segment. This can be used to compute axial current given the voltage at two adjacent points. If there is no parent the "infinite" resistance returned is 1e30.

File: LFPy/cell.py
Comment (espenhgn) [inline LFPy/cell.py]: We discussed already that this function does not belong here, but in the file with different dipole forward models for potentials and magnetic fields. For one it does not use any attributes of `self` and should not be a class method.

File: LFPy/cell.py
Comment (espenhgn) [inline LFPy/cell.py]: Also here the description is a bit imprecise, it is the axial resistance from the parent segment, not the midpoint of the parent segment, see the description of `neuron.h.ri`: 

```
Return the resistance (in megohms) between the center of the segment containing x 
and its parent segment. This can be used to compute axial current 
given the voltage at two adjacent points. If there is no parent 
the "infinite" resistance returned is 1e30. ```

Please revise to reflect this.

File: LFPy/cell.py
Comment (espenhgn) [inline LFPy/cell.py]: I think it may have been intentional but I propose that the returned arguments are shaped `(2*self.totnsegs, ...)`. Now the returned arguments are always shaped `(2*self.totnsegs-2, ...)`. I know the start-to-mid point axial current of soma section would normally be 0, as well as mid-to-end axial current of terminating segments will be zero. If number of currents are 2*self.totnsegs then one should be able to unambigously access axial currents of a segment, granted that the order of axial currents and cell segments is similar. 

Also the description of returned argument `i_axial` is confusing.

File: LFPy/cell.py
Comment (espenhgn) [inline LFPy/cell.py]: I think also this function should be moved to `eegmegcalc.py`as it's not using any attributes in the class instance.

**Comment by solveignaess** [comment]
I agree.  I'll take a look at it!

2017-02-23 17:15 GMT+01:00 Espen Hagen <notifications@github.com>:

**Comment by solveignaess** [comment]
I'm not sure I understand what you mean here. The NEURON documentation is
not clear on this point, since neuron.h.ri(x) gives different results
depending on whether the segment is a bottom segment or not. (Explanation
in thesis, page 20-23.) Let me know if I should give a more detailed
explanation in the code.

2017-02-23 17:24 GMT+01:00 Espen Hagen <notifications@github.com>:

**Comment by espenhgn** [comment]
I agree NEURON’s description of this `h.ri` function is a bit confusing, but it seems as though it is correct.
A parent segment does not appear to be a requirement per se, for example for a one-compartment model I obtain
```
In [1]: import neuron
NEURON -- VERSION 7.5 (1465:e5d2f67b802e) 2016-10-16
Duke, Yale, and the BlueBrain Project -- Copyright 1984-2016
See http://neuron.yale.edu/neuron/credits


In [2]: soma = neuron.h.Section()

In [3]: neuron.h.ri(soma(0.5).x)
Out[3]: 9.014535976724952e-05

In [4]: neuron.h.ri(soma(0.0).x)
Out[4]: 1e+30
```
Extending this to two compartments:
```
In [6]: dend = neuron.h.Section()

In [7]: dend.connect(soma, 1, 0)
Out[7]: __nrnsec_0x7ff686871460

In [8]: neuron.h.topology()

|-|       __nrnsec_0x7ff686871400(0-1)
   `|       __nrnsec_0x7ff686871460(0-1)

Out[8]: 1.0

In [9]: neuron.h.ri(dend(0.5).x)
Out[9]: 9.014535976724952e-05
```

So one always get a valid resistance value from the midpoint to its 0-end, independent of whether or not there is a connected segment there. Maybe the confusion is what is meant by a “parent”; it is not a parent segment but a node in the equivalent electric circuit, connecting one or several intracellular resistances. Then also the soma above has a “parent”

File: LFPy/eegmegcalc.py
Comment (espenhgn) [inline LFPy/eegmegcalc.py]: I think a more appropriate name for this class would be ```FourSphereVolumeConductor``` (then we later on add ```OneSphereVolumeConductor``` and ```InfiniteVolumeConductor```). I'm not sure if these classes may benefit from some class inheritance structure yet, so let's keep these separated. 

Class definitions should be inherited from the base ```object``` class, like ```class FourSphereVolumeConductor(object):```

Can you also add numpydoc style "Parameters" (types, units, description) and "Example" blocks below such that  class usage becomes clear?

File: LFPy/eegmegcalc.py
Comment (espenhgn) [inline LFPy/eegmegcalc.py]: I think such commented out code can be removed (unless you intend to put the code back).

File: LFPy/eegmegcalc.py
Comment (espenhgn) [inline LFPy/eegmegcalc.py]: I think some of these class methods are probably only meant to be used internally by the above methods. If so, can you make them non-public, i.e., precede the method names by an underscore?

File: LFPy/cell.py
Comment (espenhgn) [inline LFPy/cell.py]: remove commented out codes

**Review by espenhgn — CHANGES_REQUESTED** [review]
Very nice so far. Just a few minor remarks in this review.

**Comment by espenhgn** [comment]
Saw a few places that the current dipole moment is in units of [1E-15 mA], but this should in LFPy units be [nA*µm] = [1E-15 Am] (which is unambiguous)

File: LFPy/cell.py
Comment (espenhgn) [inline LFPy/cell.py:2073]: This docstring appears to be copy-pasted from `get_dict_of_children_idx`. Can you update accordingly?

File: LFPy/eegmegcalc.py
Comment (espenhgn) [inline LFPy/eegmegcalc.py]: I think we should use a simpler implementation and definition of the current-dipole moment potential here, also less ambiguous and likely faster to compute with `np.dot()`:

Electric potential of current-dipole moment in infinite, homogeneous and linear extracellular medium:
	\begin{equation}
	\phi{(\mathbf{R})} = \frac{\mathbf{p} \cdot \mathbf{R}}{4 \pi \sigma R^3}~,
	\end{equation}
	where $\mathbf{R}=\mathbf{r}-\mathbf{r}'$ is the vector between dipole location $\mathbf{r}'$ and measurement location $\mathbf{r}$, and $R=|\mathbf{R}|$.

**Review by espenhgn — CHANGES_REQUESTED** [review]
@solveignaess  Please merge in https://github.com/solveignaess/LFPy/pull/2
and let's discuss how to proceed. I think the docstrings need to be double checked for correctness of units etc. As I merged PR #27  in branch LFPy/LFPy:dev, you may also have to pull and merge that branch before I can merge.

File: LFPy/cell.py
Comment (solveignaess) [reply LFPy/cell.py:2073]: Forgot this one! Fixing it now.

**Comment by espenhgn** [comment]
@solveignaess You can make the failing tests work by setting dt = 2**-4. We won't support dt=0.1 when tstartms < 0.

**Comment by espenhgn** [comment]
@solveignaess Looks good! I'll merge unless you have more to add.

**Comment by solveignaess** [comment]
Ok! I think we're good.=)

**Comment by espenhgn** [comment]
@solveignaess Merged! Fixes several points in Issue #7

**Pull request merged**
